Competition Changes

jordan-powers opened the pull request:
Judging by my previous experiences with rushed competition changes, I assume this probably needs some rework.

Jedralin committed:
Limelight  WIP

Jedralin committed:
Limelight

Jedralin committed:
Merge remote-tracking branch 'origin/Limelight' into Limelight

Jedralin committed:
Limelight Update

Jedralin committed:
Revert "Limelight Update"

This reverts commit 006ab221d4aa983fb2dc91b286567ae8c23c8012.

Jedralin committed:
Stash Rem's changes

Safely store Rem's WIP work so I can checkout colorTesting

FireShredder24 committed:
Merge branch 'master' into Limelight

FireShredder24 committed:
Merge branch 'master' into Limelight

FireShredder24 committed:
Wip afternoon Friday

FireShredder24 committed:
Merge branch 'Limelight' into Comp

FireShredder24 committed:
wip fri evening

FireShredder24 committed:
WIP Friday night

jordan-powers committed:
Improve neatness and legibility

jordan-powers committed:
Fix intake

Only one motor was being stopped in idle().
Also, idle() wasn't setting the value of lastPower, so the ramping
logic always thought the motor was running at full speed and thus never
clamped the acceleration.

jordan-powers committed:
Added reverse intake button

Had to switch to RunCommands so that the buttons could interrupt each
other. If we're just using InstantCommands they interact in a strange
way.

jordan-powers committed:
Flesh out purge command

jordan-powers committed:
Intake start raised

jordan-powers committed:
Rework Shooter Command

jordan-powers committed:
Add pid boolean to ShooterSubsystem

FireShredder24 committed:
wip

FireShredder24 committed:
Disable LLight while not shooting

FireShredder24 committed:
WIP Sat. night

FireShredder24 committed:
Fix Limelight

FireShredder24 committed:
Merge branch 'Comp-Rework' into Comp

FireShredder24 committed:
Post-comp work

Fix shooter/hood behavior.
Control limelight leds.
Get PID working for shooter NEOs.
Derive shooter velocity from shooter position.
Add shooter PID setpoint pref.

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -44,49 +53,79 @@
+  public static final ShuffleboardTab matchTab = Shuffleboard.getTab("Match");
+  public static final ShuffleboardTab outreachTab = Shuffleboard.getTab("Outreach");
+  public static final ShuffleboardTab testTab = Shuffleboard.getTab("Test Tune");
+  public static final ShuffleboardTab limelightTab = Shuffleboard.getTab("Limelight");
```
Should all be private non-static.  The composition rules are clearly moving away from static usage.

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -44,49 +53,79 @@
+  // .andThen(new RunCommand(shooterSubsystem::shoot, shooterSubsystem)
+  // .alongWith(new RunCommand(shooterHoodSubsystem::deployHood,
+  // shooterHoodSubsystem)));
+
```
cleanup

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -109,23 +150,34 @@ public RobotContainer() {
-    // that.
-    purge.whenPressed(new InstantCommand(shooterSubsystem::purge, shooterSubsystem, shooterHoodSubsystem));
+    // shooterShoot.whenPressed(new InstantCommand(shooterSubsystem::shoot,
+    // shooterSubsystem, shooterHoodSubsystem));
```
cleanup

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -134,8 +186,7 @@ private void configureButtonBindings() {
   public Command getAutonomousCommand() {
-    // An ExampleCommand will run in autonomous
-    return autonomousCommand;
+    return shootFromLine;
```
Could be better.  Hardcoding the selection is bad.  SendableChooser is best.

jordan-powers commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -20,14 +20,14 @@
-  public static final int FALCON_DRIVE_LEFT_REAR_CAN_ADDR = 2;
-  public static final int FALCON_DRIVE_RIGHT_FRONT_CAN_ADDR = 3;
-  public static final int FALCON_DRIVE_RIGHT_REAR_CAN_ADDR = 4;
+  public static final int FALCON_DRIVE_LEFT_FRONT_CAN_ADDR = 3;
+  public static final int FALCON_DRIVE_LEFT_REAR_CAN_ADDR = 4;
+  public static final int FALCON_DRIVE_RIGHT_FRONT_CAN_ADDR = 1;
+  public static final int FALCON_DRIVE_RIGHT_REAR_CAN_ADDR = 2;
```
Please update the Robot System Chart.

terbos90803 commented on src/main/java/frc/robot/commands/ShootFromLine.java:
```diff
@@ -0,0 +1,52 @@
+      m_shooter.idle();
+    } else {
+      m_drive.stop();
+    }
```
There are much better ways of doing this sequencing.  That was a big basis of the Command framework.  Each of these steps can be packaged as a smaller Command (e.g. as a RunCommand) and then sequenced by chaining them together.  It's much more readable and maintainable.

terbos90803 commented on src/main/java/frc/robot/commands/ShootFromWall.java:
```diff
@@ -0,0 +1,49 @@
+      m_shooter.shoot();
+    } else if (time > 1) {
+      m_shooter.idle();
+    }
```
Rework as sequenced commands.

terbos90803 commented on src/main/java/frc/robot/controllers/ControllerBase.java:
```diff
@@ -20,9 +20,20 @@ public double getMoveRequest() {
   }
 
+  public double getShootPowerCells() {
+    return Utils.deadzone(xbox.getTriggerAxis(Hand.kRight), 0.2);
```
Why is this still here?  I thought we moved this to a TriggerButton

terbos90803 commented on src/main/java/frc/robot/controllers/ControllerBase.java:
```diff
@@ -20,9 +20,20 @@ public double getMoveRequest() {
     return turnRequest;
   }
 
+  public boolean getIncHoodPos() {
+    return xbox.getPOV() == 180;
+  }
+
+  public boolean getDecHoodPos() {
+    return xbox.getPOV() == 0;
+  }
```
Probably should be packaged differently.  Not positive about that just yet...  Argue why it should be here.

jordan-powers replied:
It feels confusing that half of the inputs are defined in the RobotContainer and half of the inputs are defined in the DriveController.
This is why in previous years we didn't use JoystickButtons, instead using [BooleanTriggers](https://github.com/momentumfrc/2019DeepSpace/blob/1c7e3cb6882befb5f5d0f91fd71fd124007be35b/2019RobotCodeDeepSpace/src/main/java/frc/robot/OI.java#L54-L60) linked to the DriveController class.

jordan-powers commented on src/main/java/frc/robot/commands/AutonDriveCommand.java:
```diff
@@ -27,6 +36,25 @@ public void initialize() {
+    double distance;
+
+    if (data.valid()) {
+      turnRequest = Utils.map(data.xCoord(), -Limelight.RANGE_X, Limelight.RANGE_X, -0.25, 0.25);
```
0.25 is a magic number.

Also, I highly recommend you use the [trajectory generator](https://docs.wpilib.org/en/latest/docs/software/examples-tutorials/trajectory-tutorial/creating-following-trajectory.html#generating-the-trajectory)

jordan-powers commented on src/main/java/frc/robot/commands/AutonDriveCommand.java:
```diff
@@ -27,6 +36,25 @@ public void initialize() {
+      turnRequest = Utils.map(data.xCoord(), -Limelight.RANGE_X, Limelight.RANGE_X, -0.25, 0.25);
+      moveRequest = 0; // Utils.map(data.dist(), -Limelight.RANGE_Y, Limelight.RANGE_Y, -1.0, 1.0);
+      distance = data.dist();
+      met = false; // data.targetMet();
```
What's the point of the `met` variable if it's always false?

terbos90803 commented on src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:
```diff
@@ -86,6 +86,7 @@ public FalconDriveSubsystem() {
     // If the robot drives backwards, flip both
     leftFront.setInverted(true);
+    leftRear.setInverted(true);
     rightFront.setInverted(false);
```
For completeness, rightRear should also be listed (or rightFront removed) to avoid confusion why imbalanced.

jordan-powers commented on src/main/java/frc/robot/commands/DriveToWall.java:
```diff
@@ -0,0 +1,39 @@
+      m_subsystem.drive(0.25, 0);
+    } else {
+      m_subsystem.stop();
+    }
```
I highly recommend you [follow a trajectory](https://docs.wpilib.org/en/latest/docs/software/examples-tutorials/trajectory-tutorial/trajectory-tutorial-overview.html)

jordan-powers commented on src/main/java/frc/robot/commands/ShootFromLine.java:
```diff
@@ -0,0 +1,52 @@
+  public void execute() {
+    double time = Timer.getMatchTime();
+    m_intake.raiseIntake();
+    if (13 > time && time > 5) {
```
The `time > 5` clause is redundant

terbos90803 replied:
Not on this line.  He's trying to express 13 > time > 5

jordan-powers commented on src/main/java/frc/robot/commands/ShootFromWall.java:
```diff
@@ -0,0 +1,49 @@
+  public void execute() {
+    double time = Timer.getMatchTime();
+    if (time > 10) {
+      m_subsystem.drive(0.25, 0);
```
I highly recommend you [follow a trajectory](https://docs.wpilib.org/en/latest/docs/software/examples-tutorials/trajectory-tutorial/trajectory-tutorial-overview.html)

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -15,16 +15,18 @@
+  private final VictorSP intakeSP2;
 
   private final DoubleSolenoid intakePistonL = new DoubleSolenoid(Constants.INTAKE_PISTON_PCM_CHAN_LF_DEPLOY,
       Constants.INTAKE_PISTON_PCM_CHAN_LF_STOW);
```
There is only 1 solenoid for the intake pistons.  Why are 2 defined here?

jordan-powers commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -8,27 +8,29 @@
 
 public class ShooterHoodSubsystem extends SubsystemBase {
-  private final CANSparkMax hoodNEO = new SimmableCANSparkMax(Constants.SPARKMAX_SHOOTER_HOOD_CAN_ADDR,
+  private final CANSparkMax hoodNEO = new CANSparkMax(Constants.SPARKMAX_SHOOTER_HOOD_CAN_ADDR,
```
What did I miss? The SimmableCANSparkMax should have been a seamless replacement

terbos90803 replied:
Is there a reason why SimmableCANSparkMax doesn't work here?

terbos90803 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -40,26 +42,38 @@
+        .MakeOvercurrentMonitor(Constants.INTAKE_VICTORSP_PDP_CHAN, UNSAFE_CURRENT_LIMIT, UNSAFE_CURRENT_TIMEOUT_MS));
+
+    intakeSP.setInverted(true);
+    intakeSP2.setInverted(false);
```
Need to validate that intake motors are wired red-red/black-black and adjust these settings if necessary for minimum future confusion.

jordan-powers commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -60,27 +62,35 @@ public ShooterHoodSubsystem() {
     reliableZero = false;
     stopHood();
+
+    CommandScheduler.getInstance().registerSubsystem(this);
```
This is handled in the constructor for the supertype SubsystemBase, so you shouldn't do it here.

jordan-powers commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -60,27 +62,35 @@ public ShooterHoodSubsystem() {
   public void moveHood(double posRequest) {
     // Used for autonomous and vision-tied control of the shooter hood.
-    hoodPID.setReference(posRequest, ControlType.kSmartMotion, 0);
+    hoodPID.setReference(posRequest, ControlType.kPosition, 0);
```
Why kPosition instead of kSmartMotion?

terbos90803 replied:
kPosition is easier to get working correctly and is sufficient for a low-inertia system like the shooter hood.  It works just like a COTS servo in this mode.

jordan-powers commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
+    double dT = currTime - lastTime;
+    lastPos = currPos;
+    lastTime = currTime;
+    currVelocity = dPos / dT * 60;
```
What's wrong with `shooterEncoder.getVelocity()` ([docs](http://www.revrobotics.com/content/sw/max/sw-docs/java/com/revrobotics/CANEncoder.html#getVelocity()))?

terbos90803 replied:
getVelocity was returning the same value as getPosition.  I have a CD thread open on this topic.  https://www.chiefdelphi.com/t/sparkmax-java-encoder-getvelocity-returning-position-value/382020

jordan-powers commented on src/main/java/frc/robot/subsystems/StorageSubsystem.java:
```diff
@@ -11,8 +11,8 @@
-  private final double SAFE_SPEED = 0;
-  private final int SAFE_COOLDOWN_MS = 1000;
+  private final double SAFE_SPEED = 1;
+  private final int SAFE_COOLDOWN_MS = 0;
```
This pretty much invalidates the entire purpose of the SafeSP class. If you don't want the current monitoring, just use the VictorSP class.

jordan-powers commented on src/main/java/frc/robot/subsystems/conditioners/CurvesConditioner.java:
```diff
@@ -3,8 +3,8 @@
-  private final double move_curve = 2;
-  private final double turn_curve = 2;
+  private final double move_curve = 1;
+  private final double turn_curve = 1;
```
I assume this was at the driver's request.
What problem were you trying to solve?
This shouldn't affect wheelies, in theory it actually makes the robot harder to control.

terbos90803 replied:
This is what I said.  I definitely saw a change in the behavior after they made this change, but it could have been as much from driver practice as from a math change.  Needs more testing.

jordan-powers commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -58,7 +57,7 @@ public static double getShooterHoodSetpoint() {
 
   public static double getShooterHoodPositionTolerance() {
-    return getDouble("SHOOTER_HOOD_POSITION_TOLERANCE", SHOOTER_HOOD_POSITION_TOLERANCE);
+    return 2;
```
Umm, what's the point of using the preferences if you're just going to adjust it in code?

terbos90803 replied:
Why is this hardcoded?  Should be getting from prefs.

jordan-powers commented on src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:
```diff
@@ -153,8 +154,8 @@ public void drive(final double moveRequest, final double turnRequest) {
       leftFront.set(ControlMode.Velocity, leftETPerS);
       rightFront.set(ControlMode.Velocity, rightETPerS);
     } else {
```
You should work on getting PID working on the drive.
That's the way to prevent wheelies since you can set the max acceleration as part of the Smart Motion.

terbos90803 commented on src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,114 @@
+
+  private static final double CAMERA_ANGLE = 56.5;
+  private static final double CAMERA_HEIGHT = 23.6;
+  private static final double TARGET_HEIGHT = 84;
```
These are specific to our robot.  They should be documented and likely moved to the Constants file.

At some point, the Limelight code should be packaged as a reusable utility with appropriate configuration params.

jordan-powers commented on src/main/java/frc/robot/commands/DriveToWall.java:
```diff
@@ -0,0 +1,39 @@
+  public DriveToWall(FalconDriveSubsystem subsystem, ShooterSubsystem shooter, StorageSubsystem storage) {
+    m_subsystem = subsystem;
+    m_shooter = shooter;
+    m_storage = storage;
```
You're not requiring the subsystems you're using

jordan-powers commented on src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,114 @@
+
+  private static final double CAMERA_ANGLE = 56.5;
+  private static final double CAMERA_HEIGHT = 23.6;
+  private static final double TARGET_HEIGHT = 84;
```
Have these been calibrated?
How variable are they, should they be in MoPrefs?

terbos90803 replied:
I think RANGE_X and RANGE_Y are constants of the Limelight.  All the others are based on our robot.

terbos90803 commented on src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,114 @@
+      this.yCoord = yAngle;
+
+      double height = TARGET_HEIGHT - CAMERA_HEIGHT;
+      double slope = Math.tan(CAMERA_ANGLE + yAngle);
```
Units are not correct?  Math.tan is in radians, but those angles are clearly in degrees

jordan-powers commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -66,19 +71,29 @@
 
     this.shooterHood = shooterHood;
 
+    NetworkTableInstance.getDefault().getTable("limelight").getEntry("ledMode").setNumber(3);
```
Hmm, this [smells](https://en.wikipedia.org/wiki/Code_smell)

terbos90803 commented on src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,114 @@
+        double x = Math.abs(xCoord());
+        double y = Math.abs(yCoord());
+        double d = Math.abs(dist() - DIST_ERR);
+        if (x <= X_ERR && y <= Y_ERR && d <= TARGET_DIST) {
```
It hardly seems likely that we can match all 3 at the same time.  In fact, y and d are coupled, so only one should be tested.

terbos90803 commented on src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,114 @@
+      return validTarget;
+    }
+
+    public boolean targetMet() {
```
I question whether targetMet is living in the correct place.  I don't see an "easier" place to move it at the moment, but that might pop up at some point.

Wherever targetMet ends up living, we should consider a companion method to generate move and turn requests based on the same formulas so we can feed that directly into the drive PID.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -8,27 +8,29 @@
-  private final double kFF = 0.000156;
-  private final double kMaxOutput = 0.3;
-  private final double kMinOutput = -0.3;
+  private final double kFF = 0;
```
kFF should be tuned first, then kP, then kI if necessary.  I suspect we won't need any kI or kD for this one.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -8,27 +8,29 @@
-  private final double kMinOutput = -0.3;
+  private final double kFF = 0;
+  private final double kMaxOutput = 1.0;
+  private final double kMinOutput = -1.0;
```
We'll always have balanced outputs here, so can collapse to a single constant that's negated as needed.

jordan-powers reviewed (commented):
I can't request changes since I technically opened the PR. But I am requesting changes.

Also, I can't leave a comment on the file `getDefault.java` because it is a completely blank file which has been created for some reason.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -60,27 +62,35 @@ public ShooterHoodSubsystem() {
+   * public void lowerShotAngle() { --hoodSetpoint; }
+   */
+
   public void moveHood(double posRequest) {
```
Might change the name to setHoodPosition or something other than move

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -93,16 +103,23 @@ public void stopHood() {
-      zeroHood();
-      reliableZero = true;
+    SmartDashboard.putNumber("pose", hoodPos);
+    if (!reliableZero) {
```
Why condition checking the limit on reliableZero?  It should probably be checked constantly.  Did you encounter a specific problem around this?

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -60,27 +62,35 @@ public ShooterHoodSubsystem() {
   public void stowHood() {
     if (reliableZero)
-      hoodPID.setReference(0, ControlType.kSmartMotion, 0);
+      hoodPID.setReference(0, ControlType.kPosition, 0);
```
Not sure if we should stow at 0.  I was thinking it should be somewhere just above the limit switch so you never touch the limit switch during normal operation.
Alternately, the stow position should *always* rely on the limit switch and not a specific position.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
+
+    final boolean shooterHoodReady = shooterHood.hasReliableZero() && shooterHood.getFullyDeployed();
+    final boolean shooterWheelReady = MoPrefs.getShooterFlywheelSetpoint()
+        - leader_shooterMAXRight.getEncoder().getVelocity() < MoPrefs.getShooterFlywheelTolerance();
```
Using the wrong getVelocity and the diff needs to be Math.abs()

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
+
+  }
+
+  public void shootAuto() {
```
Is there a particular reason why shootAuto() is different than shoot()?  Seems like a maintenance problem with no benefit.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
+      }
+    } else {
+      if (enablePID) {
+        shooterPIDRight.setReference(0, ControlType.kVelocity);
```
Probably want to use the same stopMotor for both PID and non-PID cases to take advantage of coasting instead of braking.  To be clear, the if should be eliminated and the cases collapsed.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
-    shooterHood.stowHood();
-    shooterPIDRight.setReference(-MoPrefs.getShooterFlywheelIdle(), ControlType.kVelocity);
-    shooterGate.set(-MoPrefs.getShooterGateSetpoint());
+    if (enablePID) {
```
No need for PID here.  Always use dumb speed.

No need for -1 * either, just use -
-MoPrefs.getShooterFlywheelIdle()

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
+    // NetworkTableInstance.getDefault().getTable("limelight").getEntry("ledMode").setNumber(3);
+
+    if (enablePID) {
+      shooterPIDRight.setReference(MoPrefs.getShooterPIDSetpoint(), ControlType.kVelocity);
```
Opportunity to coalesce PID and non-PID setpoints.  The ranges are mappable so there should only be one constant to control both.  Otherwise a maintenance problem.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
+    double dT = currTime - lastTime;
+    lastPos = currPos;
+    lastTime = currTime;
+    currVelocity = dPos / dT * 60;
```
If this continues to be noisy, might need to add a median filter to reject bad samples.

terbos90803 commented on src/main/java/frc/robot/subsystems/StorageSubsystem.java:
```diff
@@ -22,23 +22,25 @@
 
   public void reverse() {
-    speed = -MoPrefs.getStorageSpeed();
+    speed = -1 * MoPrefs.getStorageSpeed();
```
Why -1* instead of just -?

jordan-powers replied:
Personally, I find it more legible. The `-` can be easy to miss.

terbos90803 commented on src/main/java/frc/robot/subsystems/StorageSubsystem.java:
```diff
@@ -22,23 +22,25 @@
 
   public void run() {
     speed = MoPrefs.getStorageSpeed();
+    storage.set(speed);
```
Why set class member speed?  Just call set() with the value.

terbos90803 commented on src/main/java/frc/robot/subsystems/conditioners/ReverseConditioner.java:
```diff
@@ -1,7 +1,7 @@
 
 public class ReverseConditioner implements DriveConditioner {
-  private boolean reversed = false;
+  private boolean reversed = true;
```
Why does reversed default to true?  It should default to false and the backwards robot should be fixed elsewhere.

terbos90803 commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -3,19 +3,18 @@
-   * The maximum accleration of the intake roller, in motor setpoint per scheduler
-   * cycle.
-   */
+  static final double INTAKE_ROLLER_SETPOINT = -0.125;
```
Wow, we're really slowing down the intake.  I'm thinking we chose the wrong gearboxes.

Also, this should be a positive value.  If the motors run backwards, they should be inverted in the subsystem.

terbos90803 commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -3,19 +3,18 @@
   static final double SHOOTER_FLYWHEEL_IDLE = 0.2;
-  static final double STORAGE_SPEED = 0.5;
+  static final double STORAGE_SPEED = 0.75;
+  static final double SHOOTER_FLYWHEEL_TOLERANCE = 100; // RPM
```
Might want to tighten up this tolerance.  100 RPM could make a big difference in placement.  I seem to recall that the PID control was pretty tight, so we should be able to get this down to 10ish.  Testing required.

terbos90803 commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -3,19 +3,18 @@
-  static final double STORAGE_SPEED = 0.5;
+  static final double STORAGE_SPEED = 0.75;
+  static final double SHOOTER_FLYWHEEL_TOLERANCE = 100; // RPM
+  static final double AUTO_CHOOSER = 0;
```
What's this for?  AutoChooser should be a named pulldown, not numbers in the Prefs.

terbos90803 commented on src/main/java/frc/robot/utils/MoPrefs.java:
```diff
@@ -74,10 +73,22 @@ public static double getShooterFlywheelIdle() {
 
   public static double getIntakeRollerAccRamp() {
-    return getDouble("INTAKE_ROLLER_ACC_RAMP", INTAKE_ROLLER_ACC_RAMP);
+    return 0.1;
```
same

sciencewhiz commented on src/main/java/frc/robot/utils/SimmableCANSparkMax.java:
```diff
@@ -50,6 +59,78 @@ public CANError setReference(double value, ControlType ctrl, int pidSlot, double
+        }
+        return m_encoderCount.get();
+      } else {
+        return super.getPosition();
```
Should return super.getVelocity().

terbos90803 replied:
And there it is.  Thanks so much for tracking that one down.

jordan-powers replied:
Ah, whoops. That one's on me

jordan-powers committed:
Fix SimmableCANSparkMax Encoder Velocity

terbos90803 commented on src/main/java/frc/robot/utils/SimmableCANSparkMax.java:
```diff
@@ -125,7 +125,7 @@ public double getVelocity() {
         }
         return m_encoderCount.get();
```
This seems wrong too.  Position and Velocity should be different things.

jordan-powers replied:
Looks like I just copy-pasted the code from encoderCount and updated the
string without updating any of the actual code. Oops.

jordan-powers committed:
Fix encoderVelocity

Looks like I just copy-pasted the code from encoderCount and updated the
string without updating any of the actual code. Oops.

FireShredder24 committed:
WIP on PR changes

FireShredder24 committed:
Inline Dumb Autonomous Commands

FireShredder24 committed:
Remove unused imports

jordan-powers committed:
Merge branch 'master' into Comp

terbos90803 committed:
SparkMax cleanup

Use Simmable for the shooter hood.
Use the native getVelocity for the shooter.

FireShredder24 committed:
Auto stash before merge of "Comp" and "origin/Comp"

FireShredder24 committed:
Auto stash before merge of "Comp" and "origin/Comp"

FireShredder24 committed:
Complete old PR suggestions

Read the PR from before this date for more detailed information.
Not listed on the PR:
Removed the momentum conservation feature of the shooter flywheel, which would keep running the flywheel when not actively shooting at a lower velocity to decrease ramp-up times.  This was never used and is a maintenance issue.
Removed two Prefs, SHOOTER_FLYWHEEL_SETPOINT and SHOOTER_FLYWHEEL_IDLE that were rendered useless by other changes documented on the PR.
Added a widget on the Match tab that allows the user to choose whether or not to enable PID on the shooter flywheel.

FireShredder24 committed:
Fix all code smells assigned to me

All the code smells assigned to Jordan on SonarCloud are untouched.

FireShredder24 committed:
Merge branch 'master' into Comp

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -98,33 +89,32 @@
-  private final Command shootIdleAutoCommand = new RunCommand(shooterSubsystem::idle, shooterSubsystem)
-      .alongWith(new RunCommand(shooterHoodSubsystem::stowHood, shooterHoodSubsystem));
+  private final Command shootFromLine = new ParallelCommandGroup(
+      new RunCommand(shooterSubsystem::shoot, shooterSubsystem).withTimeout(10),
```
I'm not sure I'm fully understanding the intent of this command group.  An English description would be outstanding.

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -98,33 +89,32 @@
 
-  private final ShootFromLine shootFromLine = new ShootFromLine(falconDriveSubsystem, shooterSubsystem,
-      storageSubsystem, intakeSubsystem, shooterHoodSubsystem);
+  private final Command shootFromWall = new ParallelCommandGroup(
```
Same here.  I'm not understanding the use of the ParallelCommandGroup to achieve a sequence of steps.

terbos90803 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -98,33 +89,32 @@
+  private final Command shootFromWall = new ParallelCommandGroup(
+      new RunCommand(() -> falconDriveSubsystem.drive(0.5, 0)).withTimeout(5).andThen(
+          new RunCommand(shooterSubsystem::shoot, shooterSubsystem),
+          new RunCommand(() -> shooterHoodSubsystem.setHoodPosition(60), shooterHoodSubsystem),
```
setHoodPosition(magic_number).

I still don't understand why the shooterHood is treated differently from the shooter.  There should be a hierarchy of subsystems such that if you want to shoot, you send a command to the shooter and it figures out what to do with the hood to make the shots correctly.  The top level command shouldn't need to know about both.

terbos90803 commented on src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -71,8 +75,8 @@ private LimelightData(double valid, double xAngle, double yAngle) {
-      double height = TARGET_HEIGHT - CAMERA_HEIGHT;
-      double slope = Math.tan(CAMERA_ANGLE + yAngle);
+      double height = Constants.TARGET_HEIGHT - Constants.CAMERA_HEIGHT;
+      double slope = Math.tan(Math.toRadians(Constants.CAMERA_ANGLE) + Math.toRadians(yAngle));
```
Instead of calling toRadians twice, add the angles as degrees and then convert to radians once.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -45,50 +49,45 @@
+  private static final double kD = 0;
   /**
    * The Integral Zone of the SparkMAX PIDF controller. The integral accumulator
    * will reset once it hits this value.
```
The integral accumulator will reset if the error exceeds this value.  0 is disabled.

FireShredder24 replied:
The default NEO constants taken from last year's code have the IZ as 0, so that's what I went off of.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -121,23 +120,22 @@ public ShooterSubsystem(final ShooterHoodSubsystem shooterHood) {
     // fast shooter wheel
     // run gate if both of "" are good
-
+    double dumbSetpoint = Utils.map(MoPrefs.getShooterPIDSetpoint(), -5500, 5500, -1, 1);
```
magic numbers.

Instead of "dumb", perhaps "openLoop" is a better name choice.  There may be other options, but I would prefer an accurate description versus op-ed.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -146,73 +144,29 @@ public void shoot() {
+    SmartDashboard.putNumber("Flywheel Speed", shooterEncoder.getVelocity());
     SmartDashboard.putNumber("Flywheel Position", shooterEncoder.getPosition());
+
+    NetworkTableEntry shooterPIDchooser = tab.add("Shooter PID", true).withWidget(BuiltInWidgets.kToggleSwitch)
```
This can't be in periodic().  Adding the chooser can only happen once so it should be in init

FireShredder24 committed:
Complete PR Suggestions

See PR suggestions from the night of 5/7 for changes.
Changes not listed on PR:
Set the move_curve and turn_curve of CurvesConditioner to 2 to fix a failed unit test.
Change the arrangement of shootFromLine and shootFromWall to produce the desired action.
Fully nest control of the shooter hood inside ShooterSubsystem, so that RobotContainer doesn't have to interface with both Subsystems.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -69,12 +71,14 @@ public ShooterHoodSubsystem() {
+    if (reliableZero) {
       hoodPID.setReference(MoPrefs.getShooterHoodSetpoint(), ControlType.kPosition, 0);
-    else
+      currSetpoint = MoPrefs.getShooterHoodSetpoint();
```
I'd probably swap the order of these two lines and avoid calling MoPrefs twice.  Just call once and reuse currSetpoint in the setReference call.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -141,7 +151,29 @@ public void shoot() {
+    // fast flywheel
+    // deploy hood at lower angle
+    // run gate if both of "" are good
+    double openLoopSetpoint = Utils.map(MoPrefs.getShooterPIDSetpoint(), -5500, 5500, -1, 1);
```
This pattern has appeared multiple times.  Please refactor into a single function that eliminates duplicated code.  Also do something about the magic numbers.

FireShredder24 committed:
Address PR suggestions

FireShredder24 committed:
Address current PR suggestions

terbos90803 commented on src/main/java/frc/robot/commands/AutonDriveCommand.java:
```diff
@@ -56,7 +56,7 @@ public void execute() {
     }
     System.out.format("Target Distance:%.02f\n", distance);
-    drive_subsystem.drive(moveRequest, turnRequest);
+    drive_subsystem.drive(moveRequest * 0, turnRequest);
```
Needs a comment why this is times 0.  Is it a placeholder?  Could it just be 0?  Why?

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -178,19 +160,22 @@ public void shootFromWall() {
     // reverse gate
     // reverse shooter wheel
     leader_shooterMAXRight.set(-0.2);
     shooterGate.set(-1 * MoPrefs.getShooterGateSetpoint());
```
Why is this -1 here?  Could be just a - in front of MoPrefs...

FireShredder24 replied:
I believe Jordan mentioned earlier that a "-1 *" is far more readable than just a "-".  I can definitely see myself completely missing that character in a refactor scenario.

terbos90803 reviewed (changes requested):
Getting really close.

FireShredder24 committed:
Remove redundant code

terbos90803 reviewed (approved):
I think we're there.  I'm ready to merge.
